torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -449,6 +482,98 @@ void ProcessGroupAgent::listenLoop() {
   }
 }
 
+void ProcessGroupAgent::pollTimedOutRPCs() {
+  while (!shutdown_.load()) {
+    std::chrono::milliseconds sleepTime;
+    std::unique_lock<std::mutex> lock{futureMutex_};
+    // Estimate amount of time the first future will time out in, and sleep
+    // for that long.
+    // if there are no futures or the first future's RPC timeout is set to 0
+    // (meaning no timeout), then sleep for a set "infinity" time.
+    if (futureTimeouts_.empty() ||
+        futureTimeouts_.begin()->first == INFINITE_TIMEOUT) {
+      sleepTime = INFINITE_TIMEOUT;
+    } else {
+      const auto minFutureExpirationTime = futureTimeouts_.begin()->first;
+      const auto remainingTime = getRPCRemainingTime(minFutureExpirationTime);
+      sleepTime = std::max(remainingTime, std::chrono::milliseconds(0));
+    }
+
+    if (sleepTime == INFINITE_TIMEOUT) {
+      futureTimeoutCV_.wait(lock);
+    } else {
+      futureTimeoutCV_.wait_for(lock, sleepTime);
+    }
+
+    if (shutdown_.load()) {
+      return;
+    }
+
+    const auto timedOutFutures = processTimedOutFutures();
+
+    // Do not hold the lock while marking futures completed, as markCompleted()
+    // could invoke callbacks.
+    lock.unlock();
+    for (const auto& timedOutFuture : timedOutFutures) {
+      std::ostringstream ss;
+      ss << "RPC ran for more than " << timedOutFuture.timeout_.count()
+         << " milliseconds and timed out.";
+      const auto exceptionMsg = createExceptionResponse(
+          Message({}, {}, MessageType::EXCEPTION), ss.str());
+      timedOutFuture.future_->markCompleted(exceptionMsg);
+
+      const int dst = timedOutFuture.dstRank_;
+      recvCounts_.increment(dst);
+      futureCV_.notify_all();
+    }
+  }
+}
+
+const std::vector<ProcessGroupAgent::FutureInfo> ProcessGroupAgent::
+    processTimedOutFutures() {
```

Comment:
Notice this function is mutating the 2 maps that need the mutex's protection.

I guess it would be better if this function takes a lock, which is moved from the same lock used by `futureTimeoutCV_.wait_for(lock, sleepTime);`

And at the end of the function,
```cpp
lock.unlock()
futureCV_.notify_all();
```

Reply:
Doing this in https://github.com/pytorch/pytorch/pull/30355.